### PR TITLE
[RPC] Always send at least one progress/build event

### PR DESCRIPTION
**Pull request by cwong-ocaml**
We do this for errors, so it is probably a good idea to do this for build progress as well. These can be used to get a full snapshot of the current state of the build, whether in-progress or not.

Signed-off-by: Cameron Wong <cwong@janestreet.com>

### CODE REVIEW

File: src/dune_rpc_impl/server.ml [hunk 1]
```diff
@@ -298,14 +305,45 @@ let handler (t : t Fdecl.t) : 'a Dune_rpc_server.Handler.t =
...
+            ; remaining = current_progress.number_of_rules_discovered - complete
+            }
+          in
+          let last_event =
+            match Build_system.last_event () with
+            | Some evt -> rpc_event_of_build_event evt
+            | None ->
+              (* cwong: This is technically a lie, but the alternative is
+                 minting a new [Not_started] variant that is not substantively
+                 different *)
```
Comment (rgrinberg) [inline src/dune_rpc_impl/server.ml:335]: I think I'd prefer the `Not_started` or `Waiting` as opposed to this white lie. It is not urgent though.

File: src/dune_rpc_impl/server.ml [hunk 2]
```diff
@@ -298,14 +305,45 @@ let handler (t : t Fdecl.t) : 'a Dune_rpc_server.Handler.t =
...
+          Fiber.Pool.task t.pool ~f:(fun () ->
+              let events =
+                List.map errors ~f:(fun (e : Build_system.Error.t) ->
+                    Diagnostic.Event.Add (diagnostic_of_error e))
+              in
+              Session.notification session Server_notifications.diagnostic
+                events)
+        | Build_progress ->
+          let current_progress : Build_system.Progress.t =
+            Build_system.get_current_progress ()
```
Comment (rgrinberg) [inline src/dune_rpc_impl/server.ml:321]: What does this return if the build hasn't started yet?

Comment (cwong-ocaml) [reply src/dune_rpc_impl/server.ml:321]: If I'm understanding the code correctly, this should give `0/0`, which I think is reasonable.

File: src/dune_rpc_impl/server.ml (hunk 1 above)
Comment (cwong-ocaml) [reply src/dune_rpc_impl/server.ml:335]: I've added the new event.

File: src/dune_rpc_impl/server.ml (hunk 2 above)
Comment (rgrinberg) [reply src/dune_rpc_impl/server.ml:321]: Can't say I'm a fan of this result. It should not be displayed to the user, and the user has to reverse engineer the actual meaning behind it. I don't have a good idea on how to fix this yet, so I think we can accept this workaround for now.

Comment (cwong-ocaml) [reply src/dune_rpc_impl/server.ml:321]: One alternative would be to merge `Build.Event` and `Progress` entirely, maybe something like

```
type t =
  | Waiting
  | In_progress of { done : int ; remaining : int }
  | Failed
  | Interrupted
  | Success
```

Comment (rgrinberg) [reply src/dune_rpc_impl/server.ml:321]: That sounds reasonable to me.

Comment (cwong-ocaml) [reply src/dune_rpc_impl/server.ml:321]: Do you think this is the right PR to do it in? That seems like a bit of a larger change.

Comment (rgrinberg) [reply src/dune_rpc_impl/server.ml:321]: Another PR sounds more appropriate

**Review by rgrinberg — APPROVED** [review]

**Comment by cwong-ocaml** [comment]
I canceled the macos tests after 4 hours.

**Pull request merged**